outlines/text/generate/sequence.py

```diff
@@ -203,37 +139,53 @@ def __call__(
 
         num_prompt_tokens = token_ids.shape[-1]
 
-        if samples > 1:
-            token_ids, _ = self.step(
-                rng, num_prompt_tokens, token_ids, attention_mask, samples
-            )
-            is_finished = self.is_finished(token_ids)
-
-            num_batch_dims = token_ids.ndim - 1
-            repetitions = (samples,) + (1,) * num_batch_dims
-            attention_mask = torch.tile(attention_mask, repetitions)
-            attention_mask = self.expand_attention_mask(attention_mask)
-        else:
-            batch_shape = token_ids.shape[:-1]
-            is_finished = torch.zeros(batch_shape, dtype=torch.bool, device=self.device)
+        batch_shape = token_ids.shape[:-1]
+        token_ids = torch.broadcast_to(token_ids, (samples,) + token_ids.shape)
+        attention_mask = torch.broadcast_to(
+            attention_mask, (samples,) + attention_mask.shape
+        )
+        is_finished = torch.zeros(
+            (samples,) + batch_shape, dtype=torch.bool, device=self.device
+        )
 
         while True:
             num_generated_tokens = token_ids.shape[-1] - num_prompt_tokens
             if torch.all(is_finished) or num_generated_tokens == self.max_tokens:
                 break
 
-            updated_token_ids, _ = self.step(
+            # Draw samples only for the sequences that aren't finished
+            unfinished_token_ids = token_ids[~is_finished]
+            unfinished_attention_mask = attention_mask[~is_finished]
+            unfinished_next_token_ids, _ = self.step(
                 rng,
                 num_prompt_tokens,
-                token_ids[~is_finished],
-                attention_mask[~is_finished],
+                unfinished_token_ids,
+                unfinished_attention_mask,
+            )
+
+            # Create an array for the next tokens of every sequence, including
+            # the finished ones (but pad them)
+            next_token_ids = torch.full(
+                token_ids.shape[:-1] + (1,), self.pad_token_id, device=self.device
             )
-            token_ids = self.update_token_ids(is_finished, token_ids, updated_token_ids)
+            next_token_ids[~is_finished] = unfinished_next_token_ids.squeeze(0)
+
+            token_ids = torch.concatenate([token_ids, next_token_ids], axis=-1)
+
             attention_mask = self.expand_attention_mask(attention_mask)
+
             is_finished[~is_finished] = self.is_finished(
-                updated_token_ids[:, num_prompt_tokens:]
+                token_ids[~is_finished][:, num_prompt_tokens:]
             ).flatten()
 
+        # TODO: It looks like batch shapes with more than one dimension aren't
+        # accepted by the decoder, so is it really that we support a one
+        # dimensional batch shape given by the number of prompts or
+        # (exclusively) `samples`?  If so, that would imply we can simplify the
+        # logic above and add an explicit check/`assert`.
```

Comment:
@rlouf, do we ever expect batch shapes greater than one dimension?  

We have a potential dimension given by `samples` and another given by the number of prompts.  This could be interpreted as a batch with shape `(samples, len(prompts))` or a batch wich shape `(samples * len(prompts),)` (i.e. flattened).  The tokenizer will only decode the latter (i.e. a list of lists), so really don't have a choice unless we update something to add support for more than two dimensional inputs in that decoder logic.